File: Prism/App/TranslationOrchestrator.swift
import Foundation
import AppKit

enum TranslationPresentation: Equatable {
    /// Translate and replace in place when editable; otherwise show popup (copy only).
    case replaceInPlace(sendAfter: Bool)
    /// Always show the result panel; Replace enabled when editable.
    case popup
}

/// Como a tradução foi disparada — atalho, Enter ou botão.
enum TranslationTrigger: Equatable {
    case hotkeyTranslateOnly(chord: String)
    case hotkeyTranslateAndSend(chord: String)
    case hotkeyPopup(chord: String)
    case enterKey

    var title: String {
        switch self {
        case .hotkeyTranslateOnly(let chord):
            return "atalho «Traduzir» (\(chord))"
        case .hotkeyTranslateAndSend(let chord):
            return "atalho «Traduzir e enviar» (\(chord))"
        case .hotkeyPopup(let chord):
            return "atalho «Painel» (\(chord))"
        case .enterKey:
            return "tecla Enter (modo «traduz e envia»)"
        }
    }

    var expectedOutcome: String {
        switch self {
        case .hotkeyTranslateOnly:
            return "traduz e substitui no campo, sem enviar"
        case .hotkeyTranslateAndSend, .enterKey:
            return "traduz, substitui e envia (Return)"
        case .hotkeyPopup:
            return "mostra o resultado no painel flutuante"
        }
    }
}

@MainActor
final class TranslationOrchestrator {
    private var settings: AppSettings
    private let engine: TranslationEngine
    private let textIO: FocusedTextIO
    private var isRunning = false
    /// Timestamp when `isRunning` was set to true; used for staleness detection.
    private var runStartDate: Date?
    /// Maximum time a translation can run before `isRunning` is forcibly reset.
    private static let maxRunDuration: TimeInterval = 45

    /// True while a capture/translate or a popup Replace is in flight.
    var isBusy: Bool { isRunning }

    var onStatusChange: ((AppState.Status) -> Void)?
    var pressReturnHandler: (() -> Void)?

    init(settings: AppSettings, engine: TranslationEngine, textIO: FocusedTextIO) {
        self.settings = settings
        self.engine = engine
        self.textIO = textIO
    }

    func updateSettings(_ settings: AppSettings) {
        self.settings = settings
        AppLog.debug(
            .orchestrator,
            "Configuração do tradutor atualizada — motor \(settings.engineLogDescription), Prism \(settings.isEnabled ? "ligado" : "desligado"), painel \(settings.popupModeEnabled ? "ligado" : "desligado")"
        )
    }

    /// Select (if needed) → translate → replace or show result panel.
    func translateFocusedText(
        presentation: TranslationPresentation,
        trigger: TranslationTrigger
    ) async {
        let runID = AppLog.beginRun()
        let runStarted = Date()
        defer { AppLog.endRun() }

        AppLog.info(.orchestrator, "── Tradução \(runID) iniciada ──")
        AppLog.info(.orchestrator, "Disparo: \(trigger.title) — \(trigger.expectedOutcome)")
        AppLog.info(.orchestrator, "Motor: \(settings.engineLogDescription)")

        guard settings.isEnabled else {
            AppLog.warning(
                .orchestrator,
                "Ignorado: Prism está desligado. Ligue o interruptor no menu ou em Configurações."
            )
            onStatusChange?(.error(String(localized: "Prism is off — turn it on in Settings")))
            AppLog.info(
                .orchestrator,
                "── Tradução \(runID) cancelada em \(AppLog.duration(since: runStarted)) — app desligado ──"
            )
            return
        }

        // Staleness check: if a previous run has been stuck for too long, force-reset.
        if isRunning, let start = runStartDate {
            let elapsed = Date().timeIntervalSince(start)
            if elapsed > Self.maxRunDuration {
                AppLog.error(
                    .orchestrator,
                    "Tradução anterior presa há \(Int(elapsed)) s — forçando reset para esta nova tentativa"
                )
                isRunning = false
                runStartDate = nil
            }
        }

        guard !isRunning else {
            AppLog.warning(
                .orchestrator,
                "Ignorado: outra tradução ainda está em andamento. Espere ela terminar e tente de novo."
            )
            onStatusChange?(.error(String(localized: "Previous translation still running — wait")))
            return
        }

        // Capture the currently active application so we can re-activate it before replacing text
        let targetApp = NSWorkspace.shared.frontmostApplication
        let appName = targetApp?.localizedName ?? "app desconhecido"
        AppLog.step(
            .orchestrator, 1, of: 4,
            "App em primeiro plano: \(appName) (\(targetApp?.bundleIdentifier ?? "sem bundle"), pid \(targetApp?.processIdentifier ?? -1))"
        )

        isRunning = true
        runStartDate = Date()
        /// Panel keeps the borrowed clipboard until dismiss so Copiar is not wiped.
        var holdPasteboardForPanel = false
        defer {
            isRunning = false
            runStartDate = nil
            if !holdPasteboardForPanel {
                textIO.finishPasteboardSession()
                AppLog.debug(.orchestrator, "Fim do run — sessão de clipboard encerrada")
            }
        }

        onStatusChange?(.translating)

        do {
            AppLog.step(.orchestrator, 2, of: 4, "Capturando o texto focado em \(appName)…")
            let capture = try await textIO.selectAndReadFocusedText()
            let trimmed = capture.text.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else {
                AppLog.warning(
                    .orchestrator,
                    "Nada para traduzir: o campo veio vazio (\(capture.logSummary))"
                )
                onStatusChange?(.error(String(localized: "No text in the focused field")))
                AppLog.info(
                    .orchestrator,
                    "── Tradução \(runID) cancelada em \(AppLog.duration(since: runStarted)) — texto vazio ──"
                )
                return
            }
            AppLog.info(.orchestrator, "Captura: \(capture.logSummary)")
            AppLog.info(.orchestrator, "Prévia original: \"\(AppLog.preview(trimmed))\"")

            let incomingFrom = settings.resolvedSourceLanguage(outgoing: false)
            let incomingTo = settings.incomingTargetLanguage
            let outgoingFrom = settings.resolvedOutgoingSourceLanguage
            let outgoingTo = settings.outgoingTargetLanguage
            let detected = LanguageDetector.detect(in: trimmed)
            // Discord/Electron: seleção no compose chega como só-leitura. Se o texto já
            // está no idioma de leitura, o par auto→pt é recusado pela Apple — usa o de escrita.
            var usedOutgoingPair = capture.isEditable
            if !capture.isEditable,
               let detected,
               LanguageCode.normalize(detected) == LanguageCode.normalize(incomingTo),
               LanguageCode.normalize(outgoingTo) != LanguageCode.normalize(incomingTo)
            {
                usedOutgoingPair = true
                AppLog.info(
                    .orchestrator,
                    "Texto já em \(LanguageCode.displayName(for: detected)) (idioma de leitura) — usando o par de escrita \(LanguageCode.pairLabel(from: outgoingFrom, to: outgoingTo))"
                )
            }

            var from = usedOutgoingPair ? outgoingFrom : incomingFrom
            var to = usedOutgoingPair ? outgoingTo : incomingTo
            let pairKind = usedOutgoingPair ? "texto que você escreve" : "texto que você lê"
            AppLog.step(
                .orchestrator, 3, of: 4,
                "Traduzindo \(LanguageCode.pairLabel(from: from, to: to)) via \(settings.engineLogDescription) (par: \(pairKind))"
            )
            let translateStarted = Date()
            let outcome: TranslationOutcome
            do {
                outcome = try await engine.translate(
                    trimmed,
                    from: from,
                    to: to
                )
            } catch {
                let canRetryOutgoing = !usedOutgoingPair
                    && LanguageCode.normalize(outgoingTo) != LanguageCode.normalize(incomingTo)
                    && (error as? TranslationError)?.isUnsupportedLanguagePair == true
                guard canRetryOutgoing else { throw error }
                AppLog.info(
                    .orchestrator,
                    "Par de leitura recusado — tentando o par de escrita \(LanguageCode.pairLabel(from: outgoingFrom, to: outgoingTo))"
                )
                from = outgoingFrom
                to = outgoingTo
                usedOutgoingPair = true
                outcome = try await engine.translate(
                    trimmed,
                    from: from,
                    to: to
                )
            }
            AppLog.info(
                .orchestrator,
                "Tradução pronta em \(AppLog.duration(since: translateStarted)) — \(outcome.text.count) caracteres"
                    + (outcome.detectedSourceLanguage.map {
                        ", origem detectada: \(LanguageCode.displayName(for: $0))"
                    } ?? "")
            )
            AppLog.info(.orchestrator, "Prévia traduzida: \"\(AppLog.preview(outcome.text))\"")

            let output = preserveSurroundingWhitespace(original: capture.text, translated: outcome.text)

            let showPanel: Bool
            let canReplace: Bool
            let sendAfter: Bool
            switch presentation {
            case .replaceInPlace(let send):
                showPanel = !capture.isEditable
                canReplace = false
                sendAfter = send
            case .popup:
                showPanel = true
                canReplace = capture.isEditable
                sendAfter = false
            }

            if showPanel {
                let sourceLabel: String
                if let from {
                    sourceLabel = LanguageCode.displayName(for: from)
                } else if let detected = outcome.detectedSourceLanguage {
                    sourceLabel = LanguageCode.displayName(for: detected)
                } else {
                    sourceLabel = LanguageCode.displayName(for: LanguageCode.systemLanguage)
                }
                let targetLabel = LanguageCode.displayName(for: to)
                let replaceHint = canReplace
                    ? "botão Substituir habilitado (⏎)"
                    : "só Copiar (⌘C) — campo não é editável"
                AppLog.step(
                    .orchestrator, 4, of: 4,
                    "Abrindo painel \(sourceLabel) → \(targetLabel) — \(replaceHint)"
                )
                holdPasteboardForPanel = true
                TranslationResultPanelController.shared.show(
                    .init(
                        original: capture.text,
                        translated: output,
                        canReplace: canReplace,
                        showOriginal: canReplace,
                        sourceLanguageLabel: sourceLabel,
                        targetLanguageLabel: targetLabel,
                        pairContextLabel: usedOutgoingPair
                            ? String(localized: "Text I write")
                            : String(localized: "Text I read"),
                        capture: capture,
                        targetApp: targetApp,
                        onReplace: { [weak self] capture, text, app in
                            await self?.performReplaceFromPanel(
                                capture: capture, text: text, targetApp: app)
                        },
                        onDismissWithoutReplace: { [weak self] in
                            self?.textIO.finishPasteboardSession(onlyIfUnchanged: true)
                        }
                    )
                )
                onStatusChange?(.success)
                AppLog.info(
                    .orchestrator,
                    "── Tradução \(runID) concluída em \(AppLog.duration(since: runStarted)) — painel aberto em \(appName) ──"
                )
                return
            }

            AppLog.step(
                .orchestrator, 4, of: 4,
                sendAfter
                    ? "Substituindo o texto no campo e enviando Return"
                    : "Substituindo o texto no campo (sem enviar)"
            )
            try await performReplace(capture: capture, text: output, targetApp: targetApp)

            if sendAfter {
                try await Task.sleep(nanoseconds: 25_000_000)
                AppLog.info(.orchestrator, "Enviando tecla Return (traduzir e enviar)")
                pressReturnHandler?()
            }

            onStatusChange?(.success)
            let ending = sendAfter ? "texto substituído e enviado" : "texto substituído"
            AppLog.info(
                .orchestrator,
                "── Tradução \(runID) concluída em \(AppLog.duration(since: runStarted)) — \(ending) em \(appName) ──"
            )
        } catch {
            AppLog.error(
                .orchestrator,
                "Falhou: \(error.localizedDescription)"
            )
            AppLog.debug(.orchestrator, "Erro técnico: \(String(describing: error))")
            onStatusChange?(.error(error.localizedDescription))
            AppLog.error(
                .orchestrator,
                "── Tradução \(runID) falhou em \(AppLog.duration(since: runStarted)) — \(error.localizedDescription) ──"
            )
        }
    }

    /// Replace from the popup — holds `isRunning` so a concurrent hotkey cannot
    /// capture/paste while the clipboard session is still in use.
    private func performReplaceFromPanel(
        capture: FocusedTextCapture,
        text: String,
        targetApp: NSRunningApplication?
    ) async {
        if isRunning, let start = runStartDate {
            let elapsed = Date().timeIntervalSince(start)
            if elapsed > Self.maxRunDuration {
                isRunning = false
                runStartDate = nil
            }
        }
        guard !isRunning else {
            AppLog.warning(
                .orchestrator,
                "Ignorado: substituição do painel enquanto outra tradução ainda corre"
            )
            onStatusChange?(.error(String(localized: "Previous translation still running — wait")))
            return
        }
        isRunning = true
        runStartDate = Date()
        defer {
            isRunning = false
            runStartDate = nil
            textIO.finishPasteboardSession()
        }
        do {
            try await performReplace(capture: capture, text: text, targetApp: targetApp)
            onStatusChange?(.success)
        } catch {
            AppLog.error(
                .orchestrator,
                "Não deu para substituir o texto: \(error.localizedDescription)"
            )
            AppLog.debug(.orchestrator, "Erro técnico na substituição: \(String(describing: error))")
            onStatusChange?(.error(error.localizedDescription))
        }
    }

    private func performReplace(
        capture: FocusedTextCapture,
        text: String,
        targetApp: NSRunningApplication?
    ) async throws {
        if let targetApp, !targetApp.isActive {
            AppLog.info(
                .orchestrator,
                "Reativando \(targetApp.localizedName ?? "o app de origem") (pid \(targetApp.processIdentifier)) para colar a tradução"
            )
            if #available(macOS 14.0, *) {
                targetApp.activate()
            } else {
                targetApp.activate(options: .activateIgnoringOtherApps)
            }
            try await Task.sleep(nanoseconds: 20_000_000)
        }

        AppLog.info(
            .orchestrator,
            "Substituindo seleção — \(text.count) caracteres (\(capture.logSummary))"
        )
        try await textIO.replaceSelection(capture, with: text)
        AppLog.info(.orchestrator, "Substituição concluída")
    }

    private func preserveSurroundingWhitespace(original: String, translated: String) -> String {
        let leading = original.prefix(while: { $0.isWhitespace })
        let trailing = original.reversed().prefix(while: { $0.isWhitespace }).reversed()
        return String(leading) + translated.trimmingCharacters(in: .whitespacesAndNewlines) + String(trailing)
    }
}

File: Prism.Windows/src/Prism.Core/Orchestration/TranslationOrchestrator.cs
using Prism.Core.Policy;
using Prism.Core.Settings;
using Prism.Core.TextIO;
using Prism.Core.Translation;

namespace Prism.Core.Orchestration;

public enum AppStatusKind
{
    Idle,
    Translating,
    Success,
    Error,
}

public sealed record AppStatus(AppStatusKind Kind, string? Message = null)
{
    public static AppStatus Idle { get; } = new(AppStatusKind.Idle);
    public static AppStatus Translating { get; } = new(AppStatusKind.Translating);
    public static AppStatus Success { get; } = new(AppStatusKind.Success);
    public static AppStatus Error(string message) => new(AppStatusKind.Error, message);
}

public sealed record TranslationPanelRequest(
    string Original,
    string Translated,
    bool CanReplace,
    bool ShowOriginal,
    string SourceLanguageLabel,
    string TargetLanguageLabel,
    string PairContextLabel,
    FocusedTextCapture Capture,
    string? TargetAppId);

public interface IFocusedTextService
{
    Task<FocusedTextCapture> SelectAndReadFocusedTextAsync(
        bool preferReplaceInPlace,
        bool allowSelectAllWhenEditable,
        CancellationToken cancellationToken = default);

    Task ReplaceSelectionAsync(FocusedTextCapture capture, string text, CancellationToken cancellationToken = default);

    void FinishClipboardSession(bool onlyIfUnchanged = false);

    bool IsFocusedTextEditable();
}

public interface IForegroundAppService
{
    string? GetForegroundAppId();
    string? GetForegroundAppName();
    Task ReactivateAsync(string? appId, CancellationToken cancellationToken = default);
}

public interface IKeyboardSimulator
{
    void PressReturn();
}

public interface ITranslationResultPresenter
{
    Task ShowAsync(TranslationPanelRequest request, Func<FocusedTextCapture, string, string?, Task>? onReplace = null, Action? onDismissWithoutReplace = null);
    bool IsVisible { get; }
}

/// <summary>
/// Port of macOS TranslationOrchestrator — capture → pair selection → translate → panel or replace.
/// </summary>
public sealed class TranslationOrchestrator
{
    private static readonly TimeSpan MaxRunDuration = TimeSpan.FromSeconds(45);

    private AppSettings _settings;
    private readonly TranslationEngine _engine;
    private readonly IFocusedTextService _textIo;
    private readonly IForegroundAppService _foreground;
    private readonly IKeyboardSimulator _keyboard;
    private readonly ITranslationResultPresenter? _panel;
    private readonly ILanguageDetector _detector;

    private bool _isRunning;
    private DateTime? _runStartUtc;

    public bool IsBusy => _isRunning;
    public Action<AppStatus>? OnStatusChange { get; set; }
    public Action? PressReturnHandler { get; set; }

    public TranslationOrchestrator(
        AppSettings settings,
        TranslationEngine engine,
        IFocusedTextService textIo,
        IForegroundAppService foreground,
        IKeyboardSimulator keyboard,
        ITranslationResultPresenter? panel = null,
        ILanguageDetector? detector = null)
    {
        _settings = settings;
        _engine = engine;
        _textIo = textIo;
        _foreground = foreground;
        _keyboard = keyboard;
        _panel = panel;
        _detector = detector ?? new HeuristicLanguageDetector();
    }

    public void UpdateSettings(AppSettings settings) => _settings = settings;

    public async Task TranslateFocusedTextAsync(
        TranslationPresentation presentation,
        TranslationTrigger trigger,
        CancellationToken cancellationToken = default)
    {
        if (!_settings.IsEnabled)
        {
            OnStatusChange?.Invoke(AppStatus.Error("Prism is off — turn it on in Settings"));
            return;
        }

        if (_isRunning && _runStartUtc is { } start)
        {
            if (DateTime.UtcNow - start > MaxRunDuration)
            {
                _isRunning = false;
                _runStartUtc = null;
            }
        }

        if (_isRunning)
        {
            OnStatusChange?.Invoke(AppStatus.Error("Previous translation still running — wait"));
            return;
        }

        var targetAppId = _foreground.GetForegroundAppId();
        _isRunning = true;
        _runStartUtc = DateTime.UtcNow;
        var holdClipboardForPanel = false;

        try
        {
            OnStatusChange?.Invoke(AppStatus.Translating);

            var preferReplaceInPlace = presentation.Kind == TranslationPresentationKind.ReplaceInPlace;
            var capture = await _textIo.SelectAndReadFocusedTextAsync(
                preferReplaceInPlace,
                allowSelectAllWhenEditable: true,
                cancellationToken).ConfigureAwait(false);

            var trimmed = capture.Text.Trim();
            if (string.IsNullOrEmpty(trimmed))
            {
                OnStatusChange?.Invoke(AppStatus.Error("No text in the focused field"));
                return;
            }

            var incomingFrom = _settings.ResolvedSourceLanguage(outgoing: false);
            var incomingTo = _settings.IncomingTargetLanguage;
            var outgoingFrom = _settings.ResolvedSourceLanguage(outgoing: true);
            var outgoingTo = _settings.OutgoingTargetLanguage;
            var detected = _detector.Detect(trimmed);

            var usedOutgoingPair = capture.IsEditable;
            if (!capture.IsEditable
                && detected is not null
                && LanguageCode.Normalize(detected) == LanguageCode.Normalize(incomingTo)
                && LanguageCode.Normalize(outgoingTo) != LanguageCode.Normalize(incomingTo))
            {
                usedOutgoingPair = true;
            }

            var from = usedOutgoingPair ? outgoingFrom : incomingFrom;
            var to = usedOutgoingPair ? outgoingTo : incomingTo;

            TranslationOutcome outcome;
            try
            {
                outcome = await _engine.TranslateAsync(trimmed, from, to, cancellationToken).ConfigureAwait(false);
            }
            catch (Exception ex) when (
                !usedOutgoingPair
                && LanguageCode.Normalize(outgoingTo) != LanguageCode.Normalize(incomingTo)
                && IsUnsupportedPair(ex))
            {
                from = outgoingFrom;
                to = outgoingTo;
                usedOutgoingPair = true;
                outcome = await _engine.TranslateAsync(trimmed, from, to, cancellationToken).ConfigureAwait(false);
            }

            var output = PreserveSurroundingWhitespace(capture.Text, outcome.Text);
            var action = TranslationActionPolicy.Resolve(presentation, capture);

            if (action.ShowPanel)
            {
                var sourceLabel = from is not null
                    ? LanguageCode.DisplayNameFor(from)
                    : outcome.DetectedSourceLanguage is not null
                        ? LanguageCode.DisplayNameFor(outcome.DetectedSourceLanguage)
                        : LanguageCode.DisplayNameFor(LanguageCode.SystemLanguage);
                var targetLabel = LanguageCode.DisplayNameFor(to);

                holdClipboardForPanel = true;
                if (_panel is not null)
                {
                    await _panel.ShowAsync(
                        new TranslationPanelRequest(
                            Original: capture.Text,
                            Translated: output,
                            CanReplace: action.CanReplace,
                            ShowOriginal: action.CanReplace,
                            SourceLanguageLabel: sourceLabel,
                            TargetLanguageLabel: targetLabel,
                            PairContextLabel: usedOutgoingPair ? "Text I write" : "Text I read",
                            Capture: capture,
                            TargetAppId: targetAppId),
                        onReplace: async (cap, text, appId) =>
                            await PerformReplaceFromPanelAsync(cap, text, appId, cancellationToken).ConfigureAwait(false),
                        onDismissWithoutReplace: () => _textIo.FinishClipboardSession(onlyIfUnchanged: true)).ConfigureAwait(false);
                }

                OnStatusChange?.Invoke(AppStatus.Success);
                return;
            }

            await PerformReplaceAsync(capture, output, targetAppId, cancellationToken).ConfigureAwait(false);

            if (action.SendAfter)
            {
                await Task.Delay(25, cancellationToken).ConfigureAwait(false);
                if (PressReturnHandler is not null)
                    PressReturnHandler();
                else
                    _keyboard.PressReturn();
            }

            OnStatusChange?.Invoke(AppStatus.Success);
        }
        catch (Exception ex)
        {
            OnStatusChange?.Invoke(AppStatus.Error(ex.Message));
        }
        finally
        {
            _isRunning = false;
            _runStartUtc = null;
            if (!holdClipboardForPanel)
                _textIo.FinishClipboardSession();
        }
    }

    private async Task PerformReplaceFromPanelAsync(
        FocusedTextCapture capture,
        string text,
        string? targetAppId,
        CancellationToken cancellationToken)
    {
        if (_isRunning && _runStartUtc is { } start && DateTime.UtcNow - start > MaxRunDuration)
        {
            _isRunning = false;
            _runStartUtc = null;
        }

        if (_isRunning)
        {
            OnStatusChange?.Invoke(AppStatus.Error("Previous translation still running — wait"));
            return;
        }

        _isRunning = true;
        _runStartUtc = DateTime.UtcNow;
        try
        {
            await PerformReplaceAsync(capture, text, targetAppId, cancellationToken).ConfigureAwait(false);
            OnStatusChange?.Invoke(AppStatus.Success);
        }
        catch (Exception ex)
        {
            OnStatusChange?.Invoke(AppStatus.Error(ex.Message));
        }
        finally
        {
            _isRunning = false;
            _runStartUtc = null;
            _textIo.FinishClipboardSession();
        }
    }

    private async Task PerformReplaceAsync(
        FocusedTextCapture capture,
        string text,
        string? targetAppId,
        CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(text))
            throw TranslationError.EmptyResponse();

        await _foreground.ReactivateAsync(targetAppId, cancellationToken).ConfigureAwait(false);
        await _textIo.ReplaceSelectionAsync(capture, text, cancellationToken).ConfigureAwait(false);
    }

    private static bool IsUnsupportedPair(Exception ex) =>
        ex is TranslationError te && te.IsUnsupportedLanguagePair;

    internal static string PreserveSurroundingWhitespace(string original, string translated)
    {
        var leadingLen = 0;
        while (leadingLen < original.Length && char.IsWhiteSpace(original[leadingLen]))
            leadingLen++;

        var trailingLen = 0;
        while (trailingLen < original.Length - leadingLen
               && char.IsWhiteSpace(original[original.Length - 1 - trailingLen]))
            trailingLen++;

        var leading = original[..leadingLen];
        var trailing = trailingLen > 0 ? original[^trailingLen..] : "";
        return leading + translated.Trim() + trailing;
    }
}
